### PR TITLE
chore(api-v3): support changed conversation.access-update event AR-2743

### PR DESCRIPTION
**Pull request by typfel**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

The `access_role_v2` will be renamed to `access_role` in a future version of the backend so during a transition period we will need to support both cases.

For more details see: https://wearezeta.atlassian.net/wiki/spaces/ENGINEERIN/pages/672006169/API+changes+v2+v3?focusedCommentId=684916909#comment-684916909

#### Test Coverage (Optional)

- [x] I have added automated test to this contribution

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)

_Attachments like images, videos, etc. (drag and drop in the text box)_

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: network/src/commonMain/kotlin/com/wire/kalium/network/api/base/authenticated/conversation/model/ConversationAccessInfoDTO.kt [hunk 1]
```diff
@@ -5,11 +5,18 @@ import com.wire.kalium.network.api.base.model.ConversationAccessDTO
...
+ * **Deprecation info**: Since API v3 `access_role_v2` is deprecated and will be replaced by `access_role`, but until all servers have
+ * been upgraded to have API v3 has its minimum supported version and all previously stored events have expired we need support both cases.
+ *
+ * Further info: https://wearezeta.atlassian.net/wiki/spaces/ENGINEERIN/pages/672006169/API+changes+v2+v3
+ */
 @Serializable
 data class ConversationAccessInfoDTO(
     @SerialName("access") val access: Set<ConversationAccessDTO>,
-    @SerialName("access_role_v2") val accessRole: Set<ConversationAccessRoleDTO> = ConversationAccessRoleDTO.DEFAULT_VALUE_WHEN_NULL
+    @SerialName("access_role_v2") @JsonNames("access_role") val accessRole: Set<ConversationAccessRoleDTO> = ConversationAccessRoleDTO.DEFAULT_VALUE_WHEN_NULL
```
[the remark this answers is not in the document]
Comment (MohamadJaara) [reply network/src/commonMain/kotlin/com/wire/kalium/network/api/base/authenticated/conversation/model/ConversationAccessInfoDTO.kt:19]: kotlinx serialization does support where we can change the json name only for v3 [polymorphism](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/polymorphism.md)

File: network/src/commonTest/kotlin/com/wire/kalium/api/base/authenticated/notification/AccessUpdateTest.kt [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+import com.wire.kalium.model.EventContentDTOJson
+import com.wire.kalium.network.api.base.authenticated.conversation.model.ConversationAccessInfoDTO
+import com.wire.kalium.network.api.base.authenticated.notification.EventContentDTO
+import com.wire.kalium.network.api.base.model.ConversationAccessRoleDTO
+import com.wire.kalium.network.tools.KtxSerializer
+import kotlin.test.Test
+import kotlin.test.assertEquals
+
+class AccessUpdateTest {
```
Comment (MohamadJaara) [inline network/src/commonTest/kotlin/com/wire/kalium/api/base/authenticated/notification/AccessUpdateTest.kt:11]: Maybe we can add another test for the case if legacy `access_role` and `access_role_v2` are presented in the response

File: network/src/commonMain/kotlin/com/wire/kalium/network/api/base/authenticated/conversation/model/ConversationAccessInfoDTO.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (typfel) [reply network/src/commonMain/kotlin/com/wire/kalium/network/api/base/authenticated/conversation/model/ConversationAccessInfoDTO.kt:19]: > kotlinx serialization does support [polymorphism](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/polymorphism.md) where we can change the json name only for v3

Ah cool, events are not tied to api version though so don't think it's necessary in this case. The API for events is decided when they are persisted in backend database.

> afsik API v0-2 may also send legacy access role under the same json name access_role, or the legacy has been removed completely ?

Right,  but we first look for `access_role_v2` and only if it's missing do we look for `access_role`. This is what I test in `givenPayloadWithAccessRoleAndDeprecatedAccessRoleField_whenDecoding_thenDeprecatedFieldIsPreferred()`

File: network/src/commonTest/kotlin/com/wire/kalium/api/base/authenticated/notification/AccessUpdateTest.kt (hunk 2 above)
Comment (typfel) [reply network/src/commonTest/kotlin/com/wire/kalium/api/base/authenticated/notification/AccessUpdateTest.kt:11]: If both fields are present in the response? That's I'm testing in `givenPayloadWithAccessRoleAndDeprecatedAccessRoleField_whenDecoding_thenDeprecatedFieldIsPreferred()`

Comment (MohamadJaara) [reply network/src/commonTest/kotlin/com/wire/kalium/api/base/authenticated/notification/AccessUpdateTest.kt:11]: my bad did not notice it

**Review by MohamadJaara — APPROVED** [review]
LGTM 🚀

**Pull request merged**